File: src/test/java/com/ragin/bdd/cucumbertests/hooks/ParameterizedCustomMultiParameterMatcher.java
package com.ragin.bdd.cucumbertests.hooks;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.jayway.jsonpath.JsonPath;
import com.ragin.bdd.cucumber.core.ScenarioStateContext;
import com.ragin.bdd.cucumber.matcher.BddCucumberJsonMatcher;
import net.javacrumbs.jsonunit.core.ParametrizedMatcher;
import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.json.JSONObject;

public class ParameterizedCustomMultiParameterMatcher extends BaseMatcher<Object> implements BddCucumberJsonMatcher, ParametrizedMatcher {
    private JsonParameter jsonParameter;

    @Override
    public String matcherName() {
        return "containsOneOf";
    }

    @Override
    public Class<? extends BaseMatcher<?>> matcherClass() {
        return this.getClass();
    }

    @Override
    public void setParameter(String parameter) {
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            this.jsonParameter = objectMapper.readValue(parameter, JsonParameter.class);
        } catch (Exception e) {}
    }

    @Override
    public boolean matches(final Object actual) {
        final String actualParameter = String.valueOf(actual);
        if ((isFirstValid(actualParameter) || isSecondValid(actualParameter)) &&
                ! (isFirstValid(actualParameter) && isSecondValid(actualParameter))
        ) {
            return true;
        }
        return false;
    }

    @Override
    public void describeTo(Description description) {
    }

    private boolean isFirstValid(final String actualParameter) {
        return actualParameter.equalsIgnoreCase(jsonParameter.getFirstArgument());
    }

    private boolean isSecondValid(final String actualParameter) {
        return actualParameter.equalsIgnoreCase(jsonParameter.getSecondArgument());
    }

    /**
     * Inner class for showcase
     */
    static class JsonParameter {
        private String firstArgument;
        private String secondArgument;

        public String getFirstArgument() {
            return firstArgument;
        }

        public void setFirstArgument(String firstArgument) {
            this.firstArgument = firstArgument;
        }

        public String getSecondArgument() {
            return secondArgument;
        }

        public void setSecondArgument(String secondArgument) {
            this.secondArgument = secondArgument;
        }
    }
}

File: src/test/kotlin/com/ragin/bdd/cucumbertests/hooks/ParameterizedCustomMultiParameterMatcher.kt
package com.ragin.bdd.cucumbertests.hooks

import com.ragin.bdd.cucumber.matcher.BddCucumberJsonMatcher
import com.ragin.bdd.cucumber.utils.JacksonUtils
import io.github.oshai.kotlinlogging.KotlinLogging
import net.javacrumbs.jsonunit.core.ParametrizedMatcher
import org.hamcrest.BaseMatcher
import org.hamcrest.Description
import org.junit.jupiter.api.assertNotNull

class ParameterizedCustomMultiParameterMatcher : BaseMatcher<Any>(), BddCucumberJsonMatcher, ParametrizedMatcher {
    private var jsonParameter: JsonParameter? = null

    override fun matcherName(): String {
        return "containsOneOf"
    }

    override fun matcherClass(): Class<out BaseMatcher<*>?> {
        return this.javaClass
    }

    override fun setParameter(parameter: String?) {
        assertNotNull(
            actual = parameter,
            message = "The parameter {$parameter} of 'containsOneOf' matcher cannot be null."
        )
        runCatching {
            this.jsonParameter = JacksonUtils.mapper.readValue(parameter, JsonParameter::class.java)
        }.onFailure {
            log.error { "Parameter not parsable: $parameter" }
        }
    }

    override fun matches(actual: Any): Boolean {
        val actualParameter = actual.toString()
        return (isFirstValid(actualParameter) || isSecondValid(actualParameter)) &&
                !(isFirstValid(actualParameter) && isSecondValid(actualParameter))
    }

    override fun describeTo(description: Description) = Unit

    private fun isFirstValid(actualParameter: String): Boolean {
        return actualParameter.equals(jsonParameter!!.firstArgument, ignoreCase = true)
    }

    private fun isSecondValid(actualParameter: String): Boolean {
        return actualParameter.equals(jsonParameter!!.secondArgument, ignoreCase = true)
    }

    /**
     * Inner class for showcase
     */
    internal class JsonParameter {
        var firstArgument: String? = null
        var secondArgument: String? = null
    }

    companion object {
        private val log = KotlinLogging.logger {}
    }
}
